src/passes/Inlining.cpp

```diff
@@ -298,68 +298,136 @@ struct Updater : public PostWalker<Updater> {
+      Block* branchBlock = builder->makeBlock();
+      for (Index i = 0; i < sig.params.size(); ++i) {
+        auto var = builder->addVar(getFunction(), sig.params[i]);
+        branchBlock->list.push_back(
+          builder->makeLocalSet(var, curr->operands[i]));
+        curr->operands[i] = builder->makeLocalGet(var, sig.params[i]);
+      }
+      if (auto* call = curr->template dynCast<CallIndirect>()) {
+        auto var = builder->addVar(getFunction(), Type::i32);
+        branchBlock->list.push_back(builder->makeLocalSet(var, call->target));
+        call->target = builder->makeLocalGet(var, Type::i32);
+      } else if (auto* call = curr->template dynCast<CallRef>()) {
+        assert(call->target->type.isSignature());
+        auto var = builder->addVar(getFunction(), call->target->type);
+        branchBlock->list.push_back(builder->makeLocalSet(var, call->target));
+        call->target = builder->makeLocalGet(var, call->target->type);
+      }
+      branchBlock->list.push_back(branch);
+      branchBlock->finalize(Type::unreachable);
+      replaceCurrent(branchBlock);
     } else {
-      replaceCurrent(builder->blockify(curr, builder->makeBreak(returnName)));
+      replaceCurrent(branch);
     }
+    curr->isReturn = false;
+    curr->type = sig.results;
+    returnCallInfos.push_back({curr, branch});
   }
+
   void visitCall(Call* curr) {
-    if (curr->isReturn) {
-      handleReturnCall(curr, module->getFunction(curr->target)->getResults());
-    }
+    handleReturnCall(curr, module->getFunction(curr->target)->getSig());
   }
+
   void visitCallIndirect(CallIndirect* curr) {
-    if (curr->isReturn) {
-      handleReturnCall(curr, curr->heapType.getSignature().results);
-    }
+    handleReturnCall(curr, curr->heapType.getSignature());
   }
+
   void visitCallRef(CallRef* curr) {
     Type targetType = curr->target->type;
-    if (targetType.isNull()) {
-      // We don't know what type the call should return, but we can't leave it
-      // as a potentially-invalid return_call_ref, either.
-      replaceCurrent(getDroppedChildrenAndAppend(
-        curr, *module, options, Builder(*module).makeUnreachable()));
+    if (!targetType.isSignature()) {
+      // We don't know what type the call should return, but it will also never
+      // be reached, so we don't need to do anything here.
       return;
     }
-    if (curr->isReturn) {
-      handleReturnCall(curr, targetType.getHeapType().getSignature().results);
-    }
+    handleReturnCall(curr, targetType.getHeapType().getSignature());
   }
+
   void visitLocalGet(LocalGet* curr) {
     curr->index = localMapping[curr->index];
   }
+
   void visitLocalSet(LocalSet* curr) {
     curr->index = localMapping[curr->index];
   }
+
+  void walk(Expression*& curr) {
+    PostWalker<Updater>::walk(curr);
+    if (returnCallInfos.empty()) {
+      return;
+    }
+
+    Block* body = curr->cast<Block>();
```

Comment:
Why must the function body be a block?

Reply:
This is the block we create to hold the inlined contents in the parent. We always create a block in case there are any `return`s in the inlined contents that would need to be transformed into branches out of that block.

Reply:
Are you sure you are passing the block? The inliner is crashing on this example.
```wasm
(module
 (import "env" "h" (func $h (result i32)))
 (func $g (param $x i32) (result i32)
  (if (result i32) (local.get $x)
   (then
    (return_call $h)
   )
   (else
    (i32.const 17)
   )
  )
 )
 (func (export "f") (param $x i32) (result i32)
  (call $g (local.get $x))
 )
)
```

Reply:
No, this comment was incorrect. I fixed it though, so the latest version of this branch should be good. I just tested this example locally and it worked fine.

Reply:
Oh, the fix is in #6451.

Reply:
Oops, good catch. I've moved it to the correct PR now.